meta/store.go

```diff
@@ -994,11 +1008,28 @@ func (s *Store) Authenticate(username, password string) (ui *UserInfo, err error
 			return ErrUserNotFound
 		}
 
+		// Check the local auth cache first.
+		s.mu.RLock()
+		p, ok := s.authCache[username]
+		s.mu.RUnlock()
```

Comment:
Is there a chance a "alter user" command could come in after the "runlock" bt before the "lock" below, which would delete the entry from the cache, but the the code below would shove the invalid cached value back in? If so, I think the locking needs to be of much wider scope. Alternatively you could double check that the entry exists in the cache before setting the updated cached value.